### PR TITLE
Add prerequisites for GA4 Reporting components.

### PR DESCRIPTION
**Pull request by techanvil**
## Summary

Addresses issue:

- #6559 

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/modules/analytics-4/datastore/report.js [hunk 1]
```diff
@@ -152,6 +156,104 @@ const baseSelectors = {
...
 	},
+
+	/**
+	 * Determines whether the Analytics 4 module is still gathering data.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return {boolean|undefined} Returns `true` if gathering data, otherwise `false`. Returns `undefined` while resolving.
+	 */
+	isGatheringData: createRegistrySelector( ( select ) => () => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/datastore/report.js:167]: A note to the code reviewer: Please remember, a lot of the code here is copied from existing code and modified to work with GA4. Therefore it's probably best not to treat this as a complete "root and branch" review, in that we want to remain consistent with the existing code (unless there's a compelling reason to change something).

File: assets/js/modules/analytics-4/datastore/report.js [hunk 2]
```diff
@@ -152,6 +156,104 @@ const baseSelectors = {
...
+		if ( isGatheringData === undefined ) {
+			return undefined;
+		}
+
+		// Disable reason: select needs to be called here or it will never run.
+		// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+		const report = select( MODULES_ANALYTICS_4 ).getReport( args );
+		const hasResolvedReport = select(
+			MODULES_ANALYTICS_4
+		).hasFinishedResolution( 'getReport', [ args ] );
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/datastore/report.js:240]: I believe we can return early if we know that we are still gathering data, can't we?

```js
const isGatheringData = select( MODULES_ANALYTICS_4 ).isGatheringData();
if ( isGatheringData === undefined ) {
    return undefined;
}

if ( isGatheringData === true ) {
    return true;
}

const report = select( MODULES_ANALYTICS_4 ).getReport( args );
...
```

File: assets/js/modules/analytics-4/datastore/report.js [hunk 3]
```diff
@@ -152,6 +156,104 @@ const baseSelectors = {
...
+			dimensions: [ 'date' ],
+			metrics: [ { name: 'totalUsers' } ],
+			startDate,
+			endDate,
+		};
+
+		const url = select( CORE_SITE ).getCurrentEntityURL();
+		if ( url ) {
+			args.url = url;
+		}
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/datastore/report.js:228]: Let's shuffle this a little bit. These functionality isn't needed during gathering state check, right? Then, let's move it after the gathering data status check.

File: assets/js/modules/analytics-4/utils/is-zero-report.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Checks whether the report data is empty or not.
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/is-zero-report.js:20]: I think this function checks whether the report data has zero values only, right? So, in other words the description should be something like: `Checks whether the report data has only zero values.`.

File: assets/js/modules/analytics-4/utils/is-zero-report.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object} report Report data object.
+ * @return {(boolean|undefined)} Returns undefined if in the loading state, true if the report has no data or missing data, otherwise false.
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/is-zero-report.js:25]: This function doesn't have the "loading" state. We can probably rephrase it to be something like this: `Returns TRUE if the report data has only zero values, FALSE if there is at least one non-zero value, or undefined if the report data is not resolved yet.`

File: assets/js/modules/analytics-4/utils/is-zero-report.test.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Internal dependencies.
+ */
+import { isZeroReport } from './is-zero-report';
+
+describe( 'isZeroReport', () => {
```
Comment (eugene-manuilov) [inline assets/js/modules/analytics-4/utils/is-zero-report.test.js:24]: Let's consolidate all tests in one table test that we already have here and remove other tests since they have the same logic with the one that is defined in the table test.

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @techanvil. This looks good overall, but we need to adjust a few things here. Left some comments for you. Please, take a look.

File: assets/js/modules/analytics-4/datastore/report.js (hunk 2 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/report.js:240]: Although it's correct in isolation, I don't think we should make this change. The reason being, this code is copied from the UA `isGatheringData()` with very minor changes. If and when we need to apply a change to the logic in both of these `isGatheringData()` selectors, it will be more difficult if they have diverged unnecessarily.

In fact, we have a concrete example of this in the pipeline as can be seen in https://github.com/google/site-kit-wp/pull/6565.

So, I think this is best left as-is for now.

File: assets/js/modules/analytics-4/datastore/report.js (hunk 3 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/report.js:228]: The same logic applies as in my reply to your comment on `isGatheringData()`, except in this case we don't have clear changes in the pipeline to `hasZeroData()`.

I'd suggest either leaving this as-is, or applying it both here and to the UA `hasZeroData()`. What do you think?

File: assets/js/modules/analytics-4/utils/is-zero-report.js (hunk 4 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/is-zero-report.js]: Sounds fair and although this is also copied, it's only in the JSDoc so pretty uncontentious. I've applied this change to both this and the UA version.

File: assets/js/modules/analytics-4/utils/is-zero-report.js (hunk 5 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/is-zero-report.js]: As above, I've applied this change to both this and the UA version.

File: assets/js/modules/analytics-4/utils/is-zero-report.test.js (hunk 6 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/is-zero-report.test.js:24]: Again, this is copied from the UA `isZeroReport()` tests, so it will be more maintainable to keep them the same. I'd suggest either keeping as-is, or applying the change to both UA and GA4 versions. What do you think?

File: assets/js/modules/analytics-4/datastore/report.js (hunk 3 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/datastore/report.js:228]: Yes, let's apply for both.

File: assets/js/modules/analytics-4/utils/is-zero-report.test.js (hunk 6 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/utils/is-zero-report.test.js:24]: Apply for both, please.

File: assets/js/modules/analytics-4/utils/is-zero-report.js (hunk 4 above)
Comment (eugene-manuilov) [reply assets/js/modules/analytics-4/utils/is-zero-report.js]: Thanks! 👍

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]

File: assets/js/modules/analytics-4/datastore/report.js (hunk 2 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/report.js:240]: My apologies, I thought this was in the `isGatheringData()` selector.

It seems logical to me that if we're copying the logic from one piece of code then it's easier to maintain if both of those pieces of code remain in sync, for future changes that should be applied to both of them.

However, as this is in `hasZeroData()` my comment about changes in the pipeline doesn't apply, so I'd be happy to apply this change here and to the UA version.

File: assets/js/modules/analytics-4/utils/is-zero-report.test.js (hunk 6 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/is-zero-report.test.js:24]: No problem, that's done.

File: assets/js/modules/analytics-4/datastore/report.js (hunk 3 above)
Comment (techanvil) [reply assets/js/modules/analytics-4/datastore/report.js:236]: I've applied both of the suggested changes to `hasZeroData()`.

**Review by eugene-manuilov — APPROVED** [review]
Thanks, @techanvil. It looks great now! Approved.
